[HELP] Cannot get GDB to work (Windows)

Excuse me if this was already answered, could not find it with a google search/issue search.

Hello, I'm trying to debug on my app on firebird using GDB. However, I've found my attempts unsuccessful. Do you think you can help me?

## System information
Running Windows 10 21H1 build 19043
Cygwin version 3.3.4(0.341/5/3)
GDB version 10.2
Latest firebird
Latest ndless-sdk

## Problem
I cannot get GDB to debug in firebird. Whenever I type the following:
```
arm-none-eabi-gdb -ex "target remote 127.0.0.1:3333" helloworld-sdl.elf
```
I get the following in the debugger tab of firebird:
```
Unsupported GDB cmd 'qSupported:multiprocess+;swbreak+;hwbreak+;qRelocInsn+;fork-events+;vfork-events+;exec-events+;vContSupported+;QThreadEvents+;no-resumed+;memory-tagging+'
```
Then GDB tells me:
```
Reading symbols from helloworld-sdl.elf...
Remote debugging using 127.0.0.1:3333
Remote communication error.  Target disconnected.: Broken pipe.
```

## What I've tried
1. Downgrading to an older version of GDB (Cygwin 7.9.1-1)
2. Restarting
3. Turning off anti-virus
4. Changing the port
5. Restarting without snapshot

> Great! To confirm the theory, it should be enough to edit `core/gdbstub.c` to change `set_nonblocking(socket_fd, true);` to `set_nonblocking(socket_fd, false);`.

After I run `qmake ..`, when I run `make`, I get the following error:
```
make -f Makefile.Release
make[1]: Entering directory '/cygdrive/c/Users/Majd/Documents/ndless/Firebird Emu 1.5/build'
Makefile.Release:269: *** missing separator.  Stop.
make[1]: Leaving directory '/cygdrive/c/Users/Majd/Documents/ndless/Firebird Emu 1.5/build'
make: *** [Makefile:37: release] Error 2
```
What should I do?

Contents of build: [build.zip](https://github.com/nspire-emus/firebird/files/8457304/build.zip)

I may have screwed something up along the way during the process.

When I run make this time, Makefile doesn't error. However, I get syntax errors while compiling:
```
In file included from /usr/include/qt5/QtNetwork/QtNetwork:32,
                 from /usr/include/qt5/QtQml/QtQmlDepends:4,
                 from /usr/include/qt5/QtQml/QtQml:3,
                 from ../qmlbridge.h:5,
                 from ../mainwindow.h:13,
                 from ../mainwindow.cpp:20:
/usr/include/qt5/QtNetwork/qnetworksession.h:96:32: error: expected identifier before ‘(’ token     
   96 |     QNetworkInterface interface() const;
      |                                ^
/usr/include/qt5/QtNetwork/qnetworksession.h:96:33: error: expected unqualified-id before ‘)’ token 
   96 |     QNetworkInterface interface() const;
      |                                 ^
make: *** [Makefile:894: main.o] Error 1
make: *** [Makefile:884: mainwindow.o] Error 1
In file included from /usr/include/qt5/QtNetwork/QtNetwork:32,
                 from /usr/include/qt5/QtQml/QtQmlDepends:4,
                 from /usr/include/qt5/QtQml/QtQml:3,
                 from ../qmlbridge.h:5,
                 from ../mainwindow.h:13,
                 from moc_mainwindow.cpp:9:
/usr/include/qt5/QtNetwork/qnetworksession.h:96:32: error: expected identifier before ‘(’ token     
   96 |     QNetworkInterface interface() const;
      |                                ^
/usr/include/qt5/QtNetwork/qnetworksession.h:96:33: error: expected unqualified-id before ‘)’ token 
   96 |     QNetworkInterface interface() const;
      |                                 ^
make: *** [Makefile:1248: moc_mainwindow.o] Error 1
make: Target 'first' not remade because of errors.
```

I don't know how but I solved my previous comment's issue.

But, you fix one bug, another one pops right up.

```
make: *** [Makefile:856: asmcode.o] Error 2
make: *** [Makefile:988: armsnippets_loader.o] Error 2
make: *** [Makefile:834: os-linux.o] Error 2
make: *** [Makefile:844: translate_x86_64.o] Error 2
```

This time with no description of the error.

> I may have screwed something up along the way during the process.
> 
> When I run make this time, Makefile doesn't error. However, I get syntax errors while compiling:
> 
> ```
> In file included from /usr/include/qt5/QtNetwork/QtNetwork:32,
>                  from /usr/include/qt5/QtQml/QtQmlDepends:4,
>                  from /usr/include/qt5/QtQml/QtQml:3,
>                  from ../qmlbridge.h:5,
>                  from ../mainwindow.h:13,
>                  from ../mainwindow.cpp:20:
> /usr/include/qt5/QtNetwork/qnetworksession.h:96:32: error: expected identifier before ‘(’ token     
>    96 |     QNetworkInterface interface() const;
>       |                                ^
> /usr/include/qt5/QtNetwork/qnetworksession.h:96:33: error: expected unqualified-id before ‘)’ token 
>    96 |     QNetworkInterface interface() const;
>       |                                 ^
> make: *** [Makefile:894: main.o] Error 1
> make: *** [Makefile:884: mainwindow.o] Error 1
> In file included from /usr/include/qt5/QtNetwork/QtNetwork:32,
>                  from /usr/include/qt5/QtQml/QtQmlDepends:4,
>                  from /usr/include/qt5/QtQml/QtQml:3,
>                  from ../qmlbridge.h:5,
>                  from ../mainwindow.h:13,
>                  from moc_mainwindow.cpp:9:
> /usr/include/qt5/QtNetwork/qnetworksession.h:96:32: error: expected identifier before ‘(’ token     
>    96 |     QNetworkInterface interface() const;
>       |                                ^
> /usr/include/qt5/QtNetwork/qnetworksession.h:96:33: error: expected unqualified-id before ‘)’ token 
>    96 |     QNetworkInterface interface() const;
>       |                                 ^
> make: *** [Makefile:1248: moc_mainwindow.o] Error 1
> make: Target 'first' not remade because of errors.
> ```
> 
> **(If it is important, `interface` is a macro to `__STRUCT__` which is a macro to `struct`. This is bad because there seems to be a custom definition of interface which collides with this macro. There is a `#if` check at the top and an `#undef`, but it only executes if `Q_OS_WIN` is defined which it is not because I am running on Cygwin)**

You probably shouldn't try building this in Cygwin, it'll use a different runtime than the more native windows builds. Apparently Qt isn't compatible with that already.

--- a/core/gdbstub.c
+++ b/core/gdbstub.c
@@ -125,33 +125,54 @@ static int range_translated(uint32_t range_start, uint32_t range_end) {
     return translated;
 }
 
+// returns 0 on timeout, 1 if ready (or EOF/disconnected!) and -1 on error.
+static int can_read_from_socket(int socket_fd) {
+    const int timeoutms = 100;
+
+    // There are reports that WSAPoll times out on some exceptions,
+    // so use select. Would be possible on other platforms as well,
+    // but it's limited to fds < FD_SETSIZE, unlike poll.
+#ifdef __MINGW32__
+    fd_set rfds;
+    FD_ZERO(&rfds);
+    // No need to worry about FD_SETSIZE, winsock2 uses a list instead
+    FD_SET(socket_fd, &rfds);
+    struct timeval timeout = {
+        .tv_sec = 0,
+        .tv_usec = timeoutms * 1000,
+    };
+    return select(socket_fd + 1, &rfds, NULL, NULL, &timeout);
+#else
+    struct pollfd pfd;
+    pfd.fd = socket_fd;
+    pfd.events = POLLIN;
+    return poll(&pfd, 1, timeoutms);
+#endif
+}
+
 /* Returns -1 on disconnection */
 static char get_debug_char(void) {
-    char c;
+    while(true)
+    {
+        int p = can_read_from_socket(socket_fd);
+        if(p == -1) {
+            log_socket_error("Failed to poll socket");
+            return -1;
+        }
 
-    #ifndef WIN32
-        while(true)
+        if(p) // Data available
+            break;
+
+        else // No data available
         {
-            struct pollfd pfd;
-            pfd.fd = socket_fd;
-            pfd.events = POLLIN;
-            int p = poll(&pfd, 1, 100);
-            if(p == -1) // Disconnected
+            if(exiting)
                 return -1;
 
-            if(p) // Data available
-                break;
-
-            else // No data available
-            {
-                if(exiting)
-                    return -1;
-
-                gui_do_stuff(false);
-            }
+            gui_do_stuff(false);
         }
-    #endif
+    }
 
+    char c;
     int r = recv(socket_fd, &c, 1, 0);
     if (r == -1) {
         // only for debugging - log_socket_error("Failed to recv from GDB stub socket");

